### PR TITLE
Add Arbitrary Secrets PoC flow for generating a secret

### LINKED ISSUE #9: Write protocol for creating and storing secret
This should capture the expected communications between a local client and the key servers and meet the [generate and store requirements](https://github.com/boltlabs-inc/key-mgmt-spec/blob/main/current-development-phase.md#operations-on-arbitrary-secrets). 

This protocol writeup can rely on the OPAQUE protocol as the authentication mechanism, as well use the client's output of the `export_key` from the OPAQUE handshake. Note that the AKE _MUST_ be completed before `export_key` can be used safely.

### LINKED ISSUE #42: Clarify I/O for generate and store
The client-side functionality for generating a secret and storing the result locally and remotely takes as input a `user_id` and returns a `key_id`. 

Open questions about these:
- [ ] What is the length of the user ID? Is this a global default?
- [ ] What is a key ID? What's the length?

I think it's acceptable to set both of these to be 32 bytes as a global default.

### PR DESCRIPTION
**Pull request by marsella**
Closes #9.

### CODE REVIEW

File: cryptographic_flows.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+1. The calling application initiates generation of a new secret via the local client API. They specify the type of secret (e.g. arbitrary bytes, ECDSA key pair) and provide a communication session with the key server.
+
+2. The local client forwards the request to key server. Optionally, the local client augments the request with some random bytes.
+Possible failures: the session authentication fails or times out. The networking layer times out.
+
+2. The key server selects a unique ID (among its existing set of secret IDs) and assigns it to the secret.
+
+4. The key server runs the key generation protocol: the key server generates some random bytes and combines it with the random bytes provided by the local client to create the secret.
+If the secret is an ECDSA key pair, the key server may do additional processing on the secret to make sure it meets the requirements of an ECDSA private key and computes the corresponding public key.
```
Comment (indomitableSwan) [inline cryptographic_flows.md:14]: Why don't we have local generation of the secret? There are many applications for which both local generation and caching of a secret would be appropriate, and this is more in line with our threat model. And remember that the current use case for an arbitrary secret is some action the _user_ takes, given the secret (as we are not talking about _distributed use_ yet). 

That is, our system provides "secure backup" in this case, rather than "secure use". We are building an e2e encrypted storage mechanism in this first phase of our development. The client should be sending an encrypted copy of their secret to the key server, and we should use OPAQUE to generate the encryption key, where this encryption key is known only to the client.

I think it would be helpful to clarify our design principles and threat model in a separate file, so that we can reference these more easily in the protocol specification; this should be written up in its own ticket. Using our internal design spec draft as an iteration start point would be good. However, it's worth clarifying a couple points here:
- We do not want the key server to know the secret, so the key server should not generate it. We do not fully trust the server and we want to use cryptographic protection wherever possible. This aligns with the design principle of _zero trust_.
- If we are worried about a lack of entropy on the local client device, then having the server augment the secret with randomness seems better than the other way around. This way, we have defense in depth and graceful failure: if the client has poor entropy sources, the key server may learn the secret via brute force attack, but the user is still protected from external adversaries. If the client has good entropy sources, then the key server will not be able to brute force the generated secret, and so the user is protected against an adversarial key server.

File: cryptographic_flows.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+2. The local client forwards the request to key server. Optionally, the local client augments the request with some random bytes.
+Possible failures: the session authentication fails or times out. The networking layer times out.
+
+2. The key server selects a unique ID (among its existing set of secret IDs) and assigns it to the secret.
+
+4. The key server runs the key generation protocol: the key server generates some random bytes and combines it with the random bytes provided by the local client to create the secret.
+If the secret is an ECDSA key pair, the key server may do additional processing on the secret to make sure it meets the requirements of an ECDSA private key and computes the corresponding public key. 
+
+5. The key server encrypts the secret using user-specific authentication material from the communication session.
```
Comment (indomitableSwan) [inline cryptographic_flows.md:16]: this should be done locally on the client.

File: cryptographic_flows.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+The key server stores the user ID, the key ID, the encrypted secret, and the (optional) public component in their database.
```
Comment (indomitableSwan) [inline cryptographic_flows.md:17]: We should create a separate ticket for clarifying the properties we need from the database.

File: cryptographic_flows.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+6. The key server returns the public portions of the key to the local client application, including the key ID and the public component (if any).
+
+7. The local client returns the public portions to the calling application.
+
+## Aspects not included here
+- Use permissions: there is no support for delegated keys at this point.
+- Policy engine: without a policy engine, we don't have to specify shared vs unilateral control.
+- Remote client application: the remote client can create passive secrets, but it's not incorporated here and will use a slightly different flow
+- Enclaves: The key server runs all operations on normal, untrusted hardware.
```
Comment (indomitableSwan) [inline cryptographic_flows.md:27]: I'm not sure we need to include these details here, especially because I don't think we yet know an exhaustive list. I recommend we focus on describing what we are currently building. For example, by not focusing on our current goals, we missed the fact that we want an e2e encrypted backup system from the key server in this phase.

**Review by indomitableSwan — CHANGES_REQUESTED** [review]
Thanks for this first draft! Significant revisions are required.

**Comment by indomitableSwan** [comment]
Wrote #18 and #19 to track some tasks from review comments that are not strictly part of this PR. @marsella

File: cryptographic_flows.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+The local client component can generate a secret. Using a cryptographically secure pseudo-random number generator with a random seed, generate 256 bits of output.
+
+Generate a unique tag to associate with the secret (unique among secrets stored by the local client).
+
+# Generate and store a secret locally
+The local client component can run the [generate](generate-a-secret) protocol. Then, store the tag and the secret as a pair.
+
+# Generate and store a secret remotely
+The local client must
+1. Run the [authenticate](authenticate) protocol to retrieve an export key and an open channel.
```
Comment (indomitableSwan) [inline cryptographic_flows.md:19]: - Reference the relevant issue (#4) here for authenticate, rather than including an imprecise description above.
- I really think we want to handle session abstractions separately, e.g., this isn't really what we want to happen in the implementation, is it? What if the user already has an open authenticated session and they go to generate and store a secret at the key server? It seems that we want to specify expected behavior that captures this scenario as well.

File: cryptographic_flows.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+2. Run the [generate](generate-a-secret) protocol to get a tag and a secret.
+3. Use the export key to encrypt the secret.
+4. Send the tag and encrypted secret over the open channel.
+
+The key server generates a unique secret ID for the key by generating 256 bits of randomness and hashing it with the tag to get an ID for the secret (if the ID is not unique among the key server's stored IDs, try again).
```
Comment (indomitableSwan) [inline cryptographic_flows.md:24]: What validation checks should the key server run?

Comment (indomitableSwan) [inline cryptographic_flows.md:24]: Nit: "secret ID" is a bit confusing as a name. Not sure what would be better, however.

File: entities.md [hunk 7]
```diff
@@ -1,5 +1,5 @@
-Definitions of entities involved in the key-mgmt project.
+Definitions of entities involved in the DAMS project.
```
Comment (indomitableSwan) [inline entities.md:1]: Lock Keeper.

File: entities.md [hunk 8]
```diff
@@ -1,5 +1,5 @@
...
 
 * Asset owner: the human that owns digital assets and the corresponding keys.
 * Key server: a server that participates in secure generation, storage, and use of a key (or key share).
-* User application: a middleman application that provides an interface between the asset owner and the key-mgmt client API.
\ No newline at end of file
+* User application: a middleman application that provides an interface between the asset owner and the DAMS local client API.
```
Comment (indomitableSwan) [inline entities.md:5]: Lock Keeper. Also, what does "middleman" mean? Can you clarify in the text.

File: cryptographic_flows.md (hunk 6 above)
Comment (indomitableSwan) [inline cryptographic_flows.md:24]: Do you mean the key server should generate a globally unique identifier? If so, we should say that (as well as including precisely how you want to generate the tag, i.e., which hash function).

Comment (indomitableSwan) [reply cryptographic_flows.md:24]: What hash function?

**Review by indomitableSwan — CHANGES_REQUESTED** [review]
Thank you for the rewrite! I think we're getting closer, despite my lengthy comments.

As a meta comment, I would prefer these flows to be written out with numbering. This is easier to check against implementations later (and in my opinion, to write the implementation in the first place). It also enables easier discussions on the protocol steps.

File: cryptographic_flows.md (hunk 6 above)
Comment (marsella) [reply cryptographic_flows.md:24]: Yeah, I agree. I prefer key ID but of course these are not keys yet.

Comment (marsella) [reply cryptographic_flows.md:24]: I can't think of any ID validation beyond checking uniqueness as described. Is there something else you have in mind?

File: cryptographic_flows.md (hunk 5 above)
Comment (marsella) [reply cryptographic_flows.md:19]: I changed this to just say that the session must already be open, but without the specification that it has to be step 1 here.

File: cryptographic_flows.md [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+Generate a unique tag to associate with the secret (unique among secrets stored by the local client).
+
+# Generate and store a secret locally
+The local client component can run the [generate](generate-a-secret) protocol. Then, store the tag and the secret as a pair.
+
+# Generate and store a secret remotely
+The local client must
+1. Run the [authenticate](authenticate) protocol to retrieve an export key and an open channel.
+2. Run the [generate](generate-a-secret) protocol to get a tag and a secret.
+3. Use the export key to encrypt the secret.
```
[the remark this answers is not in the document]
Comment (marsella) [reply cryptographic_flows.md:21]: Added this TODO and #27.

File: cryptographic_flows.md [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+4. Send the tag and encrypted secret over the open channel.
+
+The key server generates a unique secret ID for the key by generating 256 bits of randomness and hashing it with the tag to get an ID for the secret (if the ID is not unique among the key server's stored IDs, try again).
+Then, store the user ID (from authentication), the secret ID and encrypted secret.
```
[the remark this answers is not in the document]
Comment (marsella) [reply cryptographic_flows.md:25]: I made #28 but the description feels too vague to me. I don't know what kinds of requirements this would entail (other than, like, "not modifiable by arbitrary parties"). Do you have more specific details about what kind of requirements should be included in that issue?

[the remark this answers is not in the document]
Comment (marsella) [reply cryptographic_flows.md:25]: I also am not clear on the vision for when to write TODOs in the code vs just having an open issue to fill in details are. In this case, I don't expect to change L25 when we address #28, so I'm inclined to not put a TODO here.

File: cryptographic_flows.md [hunk 11]
```diff
@@ -0,0 +1,27 @@
+
+# Authenticate
+The local client inputs a user ID and a password and receives an "export key", which we will consider as a valid symmetric encryption key, as output. The key server receives as output some user-specific authentication material, which it can compare to its stored data to verify authentication.
+
+Both parties recieve an open channel as output, which satisfies mutual authentication, confidentiality, and integrity.
```
[the remark this answers is not in the document]
Comment (indomitableSwan) [reply cryptographic_flows.md]: The whole section.

File: cryptographic_flows.md (hunk 5 above)
Comment (indomitableSwan) [reply cryptographic_flows.md]: That sounds too vague, then. It has to be the first thing.

File: cryptographic_flows.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (indomitableSwan) [reply cryptographic_flows.md]: My thought process is that whenever something is too vague for implementation, and you don't know how to write down the details, then more work (and a TODO with an issue) is required.

In this example, L25 is very vague on what "store" means, i.e., what requirements are there? Is it OK for the server to just put all this information in any old cleartext file? What choices need to be made in order to match our threat model and general design principles?

Basically, the specification will not be complete until we have a clear definition of what it means to "store" on the server side. So L25 will change, because it should reference these requirements.

File: cryptographic_flows.md (hunk 6 above)
Comment (indomitableSwan) [reply cryptographic_flows.md]: Indeed.

File: cryptographic_flows.md [hunk 12]
```diff
@@ -1,27 +1,24 @@
...
+The local client component can generate a secret.
+1. Generate the secret: using a cryptographically secure pseudo-random number generator with a random seed, generate 256 bits of output.
+2. Generate a unique tag to associate with the secret (unique among secrets stored by the local client). A tag is 16 bytes (128 bits) of arbitrary data. It can be generated deterministically (e.g. incrementing by 1 for each new secret) or pseudorandomly.
+3. Output the tag and the secret.
 
 # Generate and store a secret locally
-The local client component can run the [generate](generate-a-secret) protocol. Then, store the tag and the secret as a pair.
+The local client component can run this operation.
+1. Execute the [generate](#generate-a-secret) protocol.
+2. Store the tag and the secret as a pair.
```
Comment (indomitableSwan) [inline cryptographic_flows.md:12]: L12 is too vague: requirements for local storage for the client should be captured as an issue and linked in a TODO comment.

File: cryptographic_flows.md [hunk 13]
```diff
@@ -1,27 +1,24 @@
...
-The local client must
-1. Run the [authenticate](authenticate) protocol to retrieve an export key and an open channel.
-2. Run the [generate](generate-a-secret) protocol to get a tag and a secret.
-3. Use the export key to encrypt the secret.
-4. Send the tag and encrypted secret over the open channel.
+To store a secret remotely, the local client must already have opened a mutually authenticated session with the key server.
 
-The key server generates a unique secret ID for the key by generating 256 bits of randomness and hashing it with the tag to get an ID for the secret (if the ID is not unique among the key server's stored IDs, try again).
-Then, store the user ID (from authentication), the secret ID and encrypted secret.
+TODO (#4 #10 #27): Describe the authentication flow and outputs precisely. The outputs should include a symmetric encryption key and a user ID.
```
Comment (indomitableSwan) [inline cryptographic_flows.md:17]: 1. The outputs for whom? 
1. Which symmetric encryption key are you referencing here? Are you saying this is the same key that will be used in L20? 
1. Are there reasons why we might consider having a layer of indirection here? i.e.,
    1. Generate a separate symmetric key that is used to encrypt our arbitrary secrets data, `backup_key`.
    1. Encrypt the `backup_key` with a symmetric key derived from the authentication flow. 
1. Does your answer to the previous question change if you think about use cases where you might want to store your arbitrary secret data _both_ locally and at the server?

File: cryptographic_flows.md [hunk 14]
```diff
@@ -1,27 +1,24 @@
...
 
-To conclude, the key server sends the secret ID to the local client.
+1. The local client runs the [generate](#generate-a-secret) protocol to get a tag and a secret.
+2. The local client encrypts the secret under the symmetric encryption key.
+3. The local client send the tag and encrypted secret over the open channel.
+4. The key server generates a globally unique identifier for the key by generating 256 bits of randomness, appending the tag to it, and computing the SHA3-256 hash to get an ID for the secret. If the ID is not unique among the key server's stored IDs, the server selects new randomness and recomputes the hash.
```
[the remark this answers is not in the document]
Comment (indomitableSwan) [reply cryptographic_flows.md:22]: In line with a previous comment of mine: what validation checks should the server run on input received from the client?

File: entities.md [hunk 15]
```diff
@@ -1,5 +1,5 @@
-Definitions of entities involved in the DAMS project.
+Definitions of entities involved in the Lock Keeper project.
 
 * Asset owner: the human that owns digital assets and the corresponding keys.
 * Key server: a server that participates in secure generation, storage, and use of a key (or key share).
-* User application: a middleman application that provides an interface between the asset owner and the DAMS local client API.
+* User application: an application that provides a direct user interface to the asset owner and calls the Lock Keeper local client API.
```
Comment (indomitableSwan) [inline entities.md:5]: I think we need to tweak this description and that this shouldn't be part of this PR. I recommend removing this text.

File: cryptographic_flows.md [hunk 16]
```diff
@@ -1,27 +1,24 @@
...
-Both parties recieve an open channel as output, which satisfies mutual authentication, confidentiality, and integrity.
-
-
 # Generate a secret
 
-The local client component can generate a secret. Using a cryptographically secure pseudo-random number generator with a random seed, generate 256 bits of output.
-
-Generate a unique tag to associate with the secret (unique among secrets stored by the local client).
+The local client component can generate a secret.
+1. Generate the secret: using a cryptographically secure pseudo-random number generator with a random seed, generate 256 bits of output.
```
Comment (indomitableSwan) [inline cryptographic_flows.md:5]: What cryptographic dependency should we rely on for the PRNG? For production of the random seed?

**Review by indomitableSwan — CHANGES_REQUESTED** [review]
Thanks for another iteration!

At the meta level, I think this is a bit too narrowly prescriptive as to the PoC's external API. e.g., it misses the possibility of wanting to generate a key and store the generated key _both_ locally and remotely. I'll also note that this PR covers #8 as well as #9, but isn't marked as such.

- TODO issue numbers do not automatically link to the issues in markdown; you have to add links manually instead.
- I put some more general design questions throughout, as well as regular comments.

**Comment by indomitableSwan** [comment]
This appears to be a duplicate issue.

File: cryptographic_flows.md [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+1. The client:
+    1. Computes `Enc(storage_key, arbitrary_key, user_id||key_id)` and sends the resulting ciphertext to the key server over the secure channel.
+    1. [Stores](#store-a-secret-locally) the ciphertext and associated data locally.
+1. The key server:
+    1. Runs a validity check on the received ciphertext (i.e., the ciphertext must be of the expected format and length).
+    1. Stores a tuple containing the received ciphertext and `user_id` in the server database.
+        - [TODO](https://github.com/boltlabs-inc/key-mgmt-spec/issues/28): Add a reference to the security requirements for server storage when complete.
+    1. Sends an ACK to the client.
+
+TODO: retries? what is the client behavior after receipt of ACK?
```
Comment (marsella) [inline cryptographic_flows.md]: Made #43 to address this

**Review by marsella — COMMENTED** [review]
I removed the key tag and added a couple missing TODOs (mentioned in comments).

**Review by indomitableSwan — APPROVED** [review]
Work on this PR became a collaboration between @marsella and myself.
